src/envs/repeated_nextto_painting.py

```diff
@@ -0,0 +1,182 @@
+        self._NextTo = Predicate("NextTo", [self._robot_type, self._obj_type],
+                                 self._NextTo_holds)
+        self._NextToBox = Predicate("NextToBox",
+                                    [self._robot_type, self._box_type],
+                                    self._NextTo_holds)
+        self._NextToShelf = Predicate("NextToShelf",
+                                      [self._robot_type, self._shelf_type],
+                                      self._NextTo_holds)
+        self._NextToNothing = Predicate("NextToNothing", [self._robot_type],
+                                        self._NextToNothing_holds)
+        # Additional Options
+        self._MoveToObj = ParameterizedOption(
+            "MoveToObj",
+            types=[self._robot_type, self._obj_type],
+            params_space=Box(self.env_lb, self.env_ub, (1, )),
+            _policy=self._Move_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._MoveToBox = ParameterizedOption(
+            "MoveToBox",
+            types=[self._robot_type, self._box_type],
+            params_space=Box(self.env_lb, self.env_ub, (1, )),
+            _policy=self._Move_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._MoveToShelf = ParameterizedOption(
+            "MoveToShelf",
+            types=[self._robot_type, self._shelf_type],
+            params_space=Box(self.env_lb, self.env_ub, (1, )),
+            _policy=self._Move_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        arr = action.arr
+        # Infer which transition function to follow
+        wash_affinity = 0 if arr[5] > 0.5 else abs(arr[5] - 0.5)
+        dry_affinity = 0 if arr[6] > 0.5 else abs(arr[6] - 0.5)
+        paint_affinity = min(abs(arr[7] - state.get(self._box, "color")),
+                             abs(arr[7] - state.get(self._shelf, "color")))
+        # NOTE: added move_affinity
+        move_affinity = sum(abs(val) for val in arr[2:])
+        affinities = [
+            (abs(1 - arr[4]), self._transition_nextto_pick_or_openlid),
+            (wash_affinity, self._transition_wash),
+            (dry_affinity, self._transition_dry),
+            (paint_affinity, self._transition_paint),
+            (abs(-1 - arr[4]), self._transition_nextto_place),
+            (move_affinity, self._transition_move),
+        ]
+        _, transition_fn = min(affinities, key=lambda item: item[0])
+        return transition_fn(state, action)
+
+    def _transition_nextto_pick_or_openlid(self, state: State,
+                                           action: Action) -> State:
+        x, y, z, _ = action.arr[:4]
+        next_state = self._transition_pick_or_openlid(state, action)
```

Comment:
I would have expected this to happen in the opposite order: first check next to, then call the parent class method. this is probably fine too, but just wondering, is there any reason that it needs to happen this way?

Reply:
oh, is it because this way saves you the trouble of figuring out which object is being picked? that makes sense.